action/protocol/rewarding/protocol_test.go

```diff
@@ -539,56 +524,104 @@ func TestStateCheckLegacy(t *testing.T) {
 }
 
 func TestMigrateValue(t *testing.T) {
-	require := require.New(t)
-	ctrl := gomock.NewController(t)
-	sm := testdb.NewMockStateManager(ctrl)
-	p := NewProtocol(
-		genesis.Default.FoundationBonusP2StartEpoch,
-		genesis.Default.FoundationBonusP2EndEpoch,
-	)
-	// put old
-	require.NoError(p.putStateV1(sm, adminKey, &admin{
-		blockReward:                big.NewInt(811),
-		epochReward:                big.NewInt(922),
-		foundationBonus:            big.NewInt(700),
-		numDelegatesForEpochReward: 118,
-	}))
-	require.NoError(p.putStateV1(sm, fundKey, &fund{
-		totalBalance:     big.NewInt(811),
-		unclaimedBalance: big.NewInt(922),
-	}))
-	require.NoError(p.putStateV1(sm, exemptKey, &exempt{
-		addrs: []address.Address{identityset.Address(0)},
-	}))
-
-	// migrate
-	require.NoError(p.migrateValueGreenland(context.Background(), sm))
-
-	// assert old (not exist)
-	_, err := p.stateV1(sm, adminKey, &admin{})
-	require.Equal(state.ErrStateNotExist, err)
-	_, err = p.stateV1(sm, fundKey, &fund{})
-	require.Equal(state.ErrStateNotExist, err)
-	_, err = p.stateV1(sm, exemptKey, &exempt{})
-	require.Equal(state.ErrStateNotExist, err)
-
-	// assert new (with correct value)
-	a := admin{}
-	_, err = p.stateV2(sm, adminKey, &a)
-	require.NoError(err)
-	require.Equal(uint64(118), a.numDelegatesForEpochReward)
-	require.Equal("811", a.blockReward.String())
-
-	f := fund{}
-	_, err = p.stateV2(sm, fundKey, &f)
-	require.NoError(err)
-	require.Equal("811", f.totalBalance.String())
-
-	e := exempt{}
-	_, err = p.stateV2(sm, exemptKey, &e)
-	require.NoError(err)
-	require.Equal(identityset.Address(0).String(), e.addrs[0].String())
+	r := require.New(t)
+
+	a1 := admin{
+		blockReward:                    big.NewInt(10),
+		epochReward:                    big.NewInt(100),
+		numDelegatesForEpochReward:     10,
+		foundationBonus:                big.NewInt(5),
+		numDelegatesForFoundationBonus: 5,
+		foundationBonusLastEpoch:       365,
+		productivityThreshold:          50,
+	}
+	f1 := fund{
+		totalBalance:     new(big.Int),
+		unclaimedBalance: new(big.Int),
+	}
+	e1 := exempt{
+		[]address.Address{identityset.Address(31)},
+	}
+	g := genesis.Default
+
+	testProtocol(t, func(t *testing.T, ctx context.Context, sm protocol.StateManager, p *Protocol) {
+		// verify v1 state
+		a := admin{}
+		_, err := p.stateV1(sm, adminKey, &a)
+		r.NoError(err)
+		r.Equal(a1, a)
+		r.False(a.hasFoundationBonusExtension())
+
+		f := fund{}
+		_, err = p.stateV1(sm, fundKey, &f)
+		r.NoError(err)
+		r.Equal(f1, f)
+
+		e := exempt{}
+		_, err = p.stateV1(sm, exemptKey, &e)
+		r.NoError(err)
+		r.Equal(e1, e)
+
+		for _, h := range []uint64{
+			g.GreenlandBlockHeight,
+			g.KamchatkaBlockHeight,
```

Comment:
this test is refactored to cover migration at both Greenland and Kamchatka height